### PR TITLE
Convert the site map to be a site map index

### PR DESCRIPTION
**Pull request by daveverwer**
This is early days but I wanted to get this documented:

- [x] `sitemap.xml` should be a site map index instead of a plain site map
- [x] Create a new static site map containing all of the top level pages like the home page, FAQ, etc…
- [x] Create a site map for every package on `/owner/repo/sitemap.xml`
- [x] Reference all package site maps in the index
- [x] Include all documentation page URLs from the data generated by `--emit-digest`

The plan is to get everything apart from the last task in the above checklist into this PR and then create a new PR for the `–emit-digest` changes that merges into this one. This is already a big PR, and that one could well be big, too.

### Go Live Plan

We should think about how we want this to go live as we want Google to see an entirely new set of URLs all at once.

- [x] Re-generate documentation for the canonical docs version so that every package that has documentation is covered by these changes
- [ ] Add some of the new sitemap files to the smoke test
- [ ] Merge PR and deploy

### CODE REVIEW

**Comment by daveverwer** [comment]
There are some problems with the URLs created by the `emit-digest` compared to the pages that we serve. I haven't looked into why they are different yet but it looks like we're including a `'` character around some special characters. Taking the digest for [SemanticVersion](https://swiftpackageindex.com/SwiftPackageIndex/SemanticVersion), we get  success on all of these URLs from the generated digest:

* [Documentation Root](https://swiftpackageindex.com/swiftpackageindex/semanticversion/0.3.5/documentation/semanticversion)
* [`SemanticVersion`](https://swiftpackageindex.com/swiftpackageindex/semanticversion/0.3.5/documentation/semanticversion/SemanticVersion)
* [`SemanticVersion/init(from:)`](https://swiftpackageindex.com/swiftpackageindex/semanticversion/0.3.5/documentation/semanticversion/SemanticVersion/init(from:))
* [`SemanticVersion/init(_:_:_:_:_:)`](https://swiftpackageindex.com/swiftpackageindex/semanticversion/0.3.5/documentation/semanticversion/SemanticVersion/init(_:_:_:_:_:))
* [`SemanticVersion/init(_:)`](https://swiftpackageindex.com/swiftpackageindex/semanticversion/0.3.5/documentation/semanticversion/SemanticVersion/init(_:))
* [`SemanticVersion/description`](https://swiftpackageindex.com/swiftpackageindex/semanticversion/0.3.5/documentation/semanticversion/SemanticVersion/description)
* [`SemanticVersion/major`](https://swiftpackageindex.com/swiftpackageindex/semanticversion/0.3.5/documentation/semanticversion/SemanticVersion/major)
* [`SemanticVersion/minor`](https://swiftpackageindex.com/swiftpackageindex/semanticversion/0.3.5/documentation/semanticversion/SemanticVersion/minor)
* [`SemanticVersion/build`](https://swiftpackageindex.com/swiftpackageindex/semanticversion/0.3.5/documentation/semanticversion/SemanticVersion/build)
* [`SemanticVersion/patch`](https://swiftpackageindex.com/swiftpackageindex/semanticversion/0.3.5/documentation/semanticversion/SemanticVersion/patch)
* [`SemanticVersion/preRelease`](https://swiftpackageindex.com/swiftpackageindex/semanticversion/0.3.5/documentation/semanticversion/SemanticVersion/preRelease)
* [`SemanticVersion/isMajorRelease`](https://swiftpackageindex.com/swiftpackageindex/semanticversion/0.3.5/documentation/semanticversion/SemanticVersion/isMajorRelease)
* [`SemanticVersion/isMinorRelease`](https://swiftpackageindex.com/swiftpackageindex/semanticversion/0.3.5/documentation/semanticversion/SemanticVersion/isMinorRelease)
* [`SemanticVersion/isPatchRelease`](https://swiftpackageindex.com/swiftpackageindex/semanticversion/0.3.5/documentation/semanticversion/SemanticVersion/isPatchRelease)
* [`SemanticVersion/isPreRelease`](https://swiftpackageindex.com/swiftpackageindex/semanticversion/0.3.5/documentation/semanticversion/SemanticVersion/isPreRelease)
* [`SemanticVersion/isInitialRelease`](https://swiftpackageindex.com/swiftpackageindex/semanticversion/0.3.5/documentation/semanticversion/SemanticVersion/isInitialRelease)
* [`SemanticVersion/isStable`](https://swiftpackageindex.com/swiftpackageindex/semanticversion/0.3.5/documentation/semanticversion/SemanticVersion/isStable)
* [`SemanticVersion/_=(_:_:)-3q5ap`](https://swiftpackageindex.com/swiftpackageindex/semanticversion/0.3.5/documentation/semanticversion/SemanticVersion/_=(_:_:)-3q5ap)
* [`SemanticVersion/_(_:_:)-4ftn7`](https://swiftpackageindex.com/swiftpackageindex/semanticversion/0.3.5/documentation/semanticversion/SemanticVersion/_(_:_:)-4ftn7)
* [`SemanticVersion/_=(_:_:)-9elz8`](https://swiftpackageindex.com/swiftpackageindex/semanticversion/0.3.5/documentation/semanticversion/SemanticVersion/_=(_:_:)-9elz8)
* [`SemanticVersion/!=(_:_:)`](https://swiftpackageindex.com/swiftpackageindex/semanticversion/0.3.5/documentation/semanticversion/SemanticVersion/!=(_:_:))
* [`SemanticVersion/_(_:_:)-1ojsm`](https://swiftpackageindex.com/swiftpackageindex/semanticversion/0.3.5/documentation/semanticversion/SemanticVersion/_(_:_:)-1ojsm)
* [`SemanticVersion/Comparable-Implementations`](https://swiftpackageindex.com/swiftpackageindex/semanticversion/0.3.5/documentation/semanticversion/SemanticVersion/Comparable-Implementations)
* [`SemanticVersion/LosslessStringConvertible-Implementations`](https://swiftpackageindex.com/swiftpackageindex/semanticversion/0.3.5/documentation/semanticversion/SemanticVersion/LosslessStringConvertible-Implementations)
* [`SemanticVersion/CustomStringConvertible-Implementations`](https://swiftpackageindex.com/swiftpackageindex/semanticversion/0.3.5/documentation/semanticversion/SemanticVersion/CustomStringConvertible-Implementations)
* [`SemanticVersion/Equatable-Implementations`](https://swiftpackageindex.com/swiftpackageindex/semanticversion/0.3.5/documentation/semanticversion/SemanticVersion/Equatable-Implementations)

Then, there are five URLs generated by the digest that fail:

* [`SemanticVersion/...(_:_:)`](https://swiftpackageindex.com/swiftpackageindex/semanticversion/0.3.5/documentation/semanticversion/SemanticVersion/...(_:_:)) we serve as [`SemanticVersion/'...(_:_:)`](https://swiftpackageindex.com/swiftpackageindex/semanticversion/0.3.5/documentation/semanticversion/SemanticVersion/'...(_:_:))
* [`SemanticVersion/...(_:)-40b95`](https://swiftpackageindex.com/swiftpackageindex/semanticversion/0.3.5/documentation/semanticversion/SemanticVersion/...(_:)-40b95) we serve as [`SemanticVersion/'...(_:)-40b95`](https://swiftpackageindex.com/swiftpackageindex/semanticversion/0.3.5/documentation/semanticversion/SemanticVersion/'...(_:)-40b95)
* [`SemanticVersion/...(_:)-bfr8`](https://swiftpackageindex.com/swiftpackageindex/semanticversion/0.3.5/documentation/semanticversion/SemanticVersion/...(_:)-bfr8) we serve as [`SemanticVersion/'...(_:)-bfr8`](https://swiftpackageindex.com/swiftpackageindex/semanticversion/0.3.5/documentation/semanticversion/SemanticVersion/'...(_:)-bfr8)
* [`SemanticVersion/.._(_:_:)`](https://swiftpackageindex.com/swiftpackageindex/semanticversion/0.3.5/documentation/semanticversion/SemanticVersion/.._(_:_:)) we serve as [`SemanticVersion/'.._(_:_:)`](https://swiftpackageindex.com/swiftpackageindex/semanticversion/0.3.5/documentation/semanticversion/SemanticVersion/'.._(_:_:))
* [`SemanticVersion/.._(_:)`](https://swiftpackageindex.com/swiftpackageindex/semanticversion/0.3.5/documentation/semanticversion/SemanticVersion/.._(_:)) we serve as [`SemanticVersion/'.._(_:)`](https://swiftpackageindex.com/swiftpackageindex/semanticversion/0.3.5/documentation/semanticversion/SemanticVersion/'.._(_:))

I haven’t looked into why we include an additional `'` in our URLs and whether it’s something we do or an artefact of DocC, but we’ll need to figure that out before it goes live.

**Comment by daveverwer** [comment]
I'd like to review this in two stages, please @finestructure.

This is enough of a PR (without implementing the final checklist item above) but I don't want to merge it until all the pieces are complete so that we show Google a full and complete sitemap all at once. We should review this now, and I'll make any changes, then I'll do the doc URLs in a separate PR that merges into this one.

File: Sources/App/Core/Extensions/SiteMap+ResponseEncodable.swift [hunk 1]
```diff
@@ -15,11 +15,18 @@
...
-extension SiteMap: ResponseEncodable {
-    public func encodeResponse(for request: Request) -> EventLoopFuture<Response> {
+extension SiteMap: AsyncResponseEncodable {
+    public func encodeResponse(for request: Request) -> Response {
         let res = Response(status: .ok, body: .init(string: self.render()))
         res.headers.add(name: "Content-Type", value: "text/xml")
-        return request.eventLoop.makeSucceededFuture(res)
+        return res
     }
 }
```
Comment (finestructure) [inline Sources/App/Core/Extensions/SiteMap+ResponseEncodable.swift:32]: Perhaps:

```swift
extension SiteMapIndex: AsyncResponseEncodable, Renderable {
    public func encodeResponse(for request: Request) async throws -> Response {
        try await encodeXMLResponse(for: request)
    }
}

extension SiteMap: AsyncResponseEncodable, Renderable {
    public func encodeResponse(for request: Request) async throws -> Response {
        try await encodeXMLResponse(for: request)
    }
}

extension AsyncResponseEncodable where Self: Renderable {
    func encodeXMLResponse(for request: Request) async throws -> Response {
        let res = Response(status: .ok, body: .init(string: self.render()))
        res.headers.add(name: "Content-Type", value: "text/xml")
        return res
    }
}
```

Also note that `AsyncResponseEncodable` protocol req for `encodeResponse` is really `async throws`. That's why this change will trigger a `try await` missing error in `SiteMapController.buildStaticPages`.

File: Sources/App/Controllers/SiteMapController.swift [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+                                                 .siteMap).absoluteURL()),
+                            .unwrap(package.lastActivityAt, { .lastmod($0) })
+                        )
+                }
+            )
+        )
+    }
+
+    static func staticPages(req: Request) async throws -> Response {
+        buildStaticPages().encodeResponse(for: req)
```
Comment (finestructure) [inline Sources/App/Controllers/SiteMapController.swift:81]: Should probably be `try await` (see below)

**Review by finestructure — APPROVED** [review]

File: Sources/App/Core/Extensions/SiteMap+ResponseEncodable.swift (hunk 1 above)
Comment (daveverwer) [reply Sources/App/Core/Extensions/SiteMap+ResponseEncodable.swift:32]: Fixed.

File: Tests/AppTests/SitemapControllerTests.swift [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+        // Validation
+        let body = try XCTUnwrap(response.body.string)
+        assertSnapshot(matching: body, as: .init(pathExtension: "xml", diffing: .lines))
+    }
+
+    @MainActor
+    func test_siteMapStaticPages() async throws {
+        // MUT
+        let req = Vapor.Request(application: app, url: "/sitemap-static-pages.xml",on: app.eventLoopGroup.next())
```
Comment (daveverwer) [inline Tests/AppTests/SitemapControllerTests.swift:48]: @finestructure I switched these for tests similar to what you suggested in https://github.com/SwiftPackageIndex/SwiftPackageIndex-Server/pull/2432#discussion_r1230522835. Unfortunately, it's not possible to pretty print as you suggested in that comment, but I think a compressed output works reasonably for sitemaps and this way is better than needing a pair of `siteMap` and `buildSiteMap` methods, as we had when you first reviewed this.

This way, we just have the one method and we test the one method.

**Comment by daveverwer** [comment]
OK, @finestructure this is ready for one final review before the second PR (which is also now ready) goes into this.

**Review by finestructure — APPROVED** [review]

File: Tests/AppTests/SitemapControllerTests.swift (hunk 3 above)
Comment (finestructure) [reply Tests/AppTests/SitemapControllerTests.swift:48]: One thing to note here: these _only_ test the returned content and _not_ whether they're wired up via routes.

I think it would be better if we test both at the same time as we normally do in these cases, via

```
app.test(.GET, "/sitemap-index.xml") { res in
    // assert status and body
}
```

Comment (finestructure) [reply Tests/AppTests/SitemapControllerTests.swift:48]: I just changed that an I'm actually getting a 404, because `/sitemap-index.xml` isn't routed - it's `/sitemap.xml` according to `SiteURL.siteMapIndex`. I'm not sure which URL it should be, so I'll adjust the test to match what's in `SiteURL` for now.

**Comment by daveverwer** [comment]
I'm going to deploy this to staging (from this branch) this afternoon and give it a test in a real environment, then I think this is ready to go.

**Comment by daveverwer** [comment]
Actually, I wonder if we should put a guard on it on staging once this test is done to make it serve empty sitemaps in that environment. We already noindex everything in staging, but why even serve a sitemap? 🤷‍♂️

**Comment by finestructure** [comment]
If there's no real downside I'd keep dev and prod the same.

**Comment by daveverwer** [comment]
I've given this a good test on staging and the only issues I found were caused by the lack of `linkable-entites.json` file in the dev bucket.

I'm going to merge and deploy it and give it one final test before asking Google to look at our sitemap again.
